### ISSUE TITLE
Prisma support

### ISSUE THREAD

**Issue by oguimbal** [oguimbal/pg-mem#27]
[Prisma](https://www.prisma.io/) support [asked here](https://www.reddit.com/r/javascript/comments/kr6ro1/pgmem_an_in_memory_postgres_db_instance_for_your/)

**Comment by nahtnam** [comment]
Would love to see this as well

**Comment by BjoernRave** [comment]
Just curious if you are still working on this lib and if this might come anyday?

**Comment by oguimbal** [comment]
I am, and if it is feasible, it will :)

**Comment by nahtnam** [comment]
I asked on slack and it seems that prisma does not use `pg`. Is it possible to mock through a connection uri?

**Comment by oguimbal** [comment]
Not a problem... pg-mem is not specific to the `pg` library. As long as Prisma is issuing PosgreSQL queries, it should be okay.

The only thing that must be done is to replace whatever lib is used to send the requests, and redirect those SQL queries to pg-mem. I havent had the time to investigate how to do that with Prisma yet, but it could be simple. For instance, [here is how](https://github.com/oguimbal/pg-mem/blob/bf2f6f7be7311a4903d7ad5b6f13ef910683899c/src/adapters.ts#L178-L188) the Slonik adapter is written, for instance (all supported libraries adapters are in this file)

It is not possible to simply use a connection uri, since pg-mem does not listen to network requests, though.

**Comment by flybayer** [comment]
It seems that integration with Prisma would be easy if pg-mem would run a server and expose a url that can be fed to prisma

**Comment by oguimbal** [comment]
Yup. I looked a bit into prisma, and it turns out that SQL queries are issued by Rust.  (I didnt know Prisma was written in Rust !).

(node) [prisma](https://github.com/prisma/prisma)  -> (rust) [prisma-engines](https://github.com/prisma/prisma-engines) -> (rust) [quaint](https://github.com/prisma/quaint) -> postgres statements.

So yes, it will be a bit invasive to try to hook quaint to execute pg-mem back in node.

I'm looking into implementing a fake postgres server. It essentially look like the other side of the mirror of [pg-protocol](https://github.com/brianc/node-postgres/tree/master/packages/pg-protocol), or more specifically for the rust version, [this](https://github.com/sfackler/rust-postgres/tree/master/postgres-protocol).... admittedly, it will be a more general solution to plug other libs/frameworks to pg-mem, but I was hoping not to implement such a low level solution 😑

**Comment by oguimbal** [comment]
For the record, to solve this, I've started working on [pg-server](https://github.com/oguimbal/pg-server), a library that can fake a postgres database. I'm working on integrating it with [pg-mem](https://github.com/oguimbal/pg-mem), which requires some refactoring to do things right. 

Once this is done, it will be possible to expose a pg-mem instance via a random network port like any regular pg database.
This will have the advantage of being compatible with any library (or any other tool, for that matter), but will add a network overhead.

I'll keep you posted when it is finished😁

The fun in all this is that pg-server is actually much more versatile that just a network interface to pg-mem... you could use it as a a db honeypot, a proxy which intercepts/inspects/modifies postgres requests on the fly (it works quite well in conjunction with [pgsql-ast-parser](https://github.com/oguimbal/pgsql-ast-parser), the syntax parser of pg-mem), or whatever !

**Comment by nahtnam** [comment]
This is super exciting! Let me know if you would like me to try a beta build of the library with my codebase (it's relatively small but has some more complex operations)

**Comment by timsuchanek** [comment]
In case you want to try out, if Prisma Client would work with `pg-mem`, you can try out our integration tests in https://github.com/prisma/prisma/tree/master/src/packages/tests
We also have a more extensive test suite written in Scala we use for Rust, but this one written in TypeScript should already get you quite far.
Normally we run them with this docker-compose file: https://github.com/prisma/prisma/blob/master/src/docker/docker-compose.yml

You should "just" need to provide the env var `TEST_POSTGRES_URI` to point to your `pg-mem` instance and then in `./src/packages/tests` you can do `yarn test:postgresql`

**Comment by mark600211** [comment]
any updates about this?

**Comment by oguimbal** [comment]
Hi, nothing new... I've worked on the refactoring I mentioned, but I've had a rough month at work, so I couldnt finish it yet.

Sorry 😶

This is definitely still in my pipe, though ! But I cant give you an ETA right now, unfortunately.

**Comment by javach** [comment]
Sorry to bump, but what's the status? Really want pg-mem integration for prisma.

**Comment by nukeop** [comment]
I'm also interested in this. Typically in integration tests in services using MongoDB, I use https://www.npmjs.com/package/mongodb-memory-server which allows connecting by URI. Having an in-memory postgres server would enable me to write quality tests for Prisma easily.

**Comment by oguimbal** [comment]
Hi !
I had very little time to work on pg-mem last year, but that's changing, and Prisma support is one of the the two most important issues for me.

A milestone has been recently reached in the refactoring that will allow this, so hopefully I will find time to get that working in the upcoming few months 🤷‍♂️

**Comment by jclaessens97** [comment]
Any updates or help required with this matter? 😊

**Comment by Asaf-S** [comment]
Really looking forward to this...

**Comment by julioxavierr** [comment]
+1

**Comment by ixartz** [comment]
+1

**Comment by ixartz** [comment]
+1, I think it's the most promising solution for in memory database working with Prisma

**Comment by grusite** [comment]
Will keep in the loop! also looking forward to this :)

**Comment by charlesilieff** [comment]
+1 that would be great

**Comment by auser** [comment]
I've *just* started working on a fork to work with Prisma, but happy to collaborate!

**Comment by Wellers0n** [comment]
+1

**Comment by tmcallister92** [comment]
+1 on this one, would be extremely useful!

**Comment by nicholasess** [comment]
+1

**Comment by jc-greenhouse** [comment]
Do we have any updates on this?

**Comment by janpio** [comment]
Small update from the Prisma side: https://github.com/prisma/prisma/issues/5419#issuecomment-1589665551

**Comment by hnrq** [comment]
Does `pg-mem` expose any kind of URL for integrating w/ Prisma? I tried looking into `pg-server` but it hasn't been updated in a while.

**Comment by SilvioLucasDev** [comment]
+1

**Comment by pitazzo** [comment]
+1

**Comment by janpio** [comment]
Another update over at the Prisma issue: https://github.com/prisma/prisma/issues/5419#issuecomment-1815922952

**Comment by oguimbal** [comment]
Hi !

Any PG protocol nerd out there that might know the answer [to this](https://github.com/prisma/prisma/issues/5419#issuecomment-2271579946) 🙂 ?

Once this is solved, me might have a real shot at closing this issue that has been there for far too long

**Comment by Francois-Esquire** [comment]
Bumping!

**Comment by Batocanin** [comment]
4 years late and still not integrated, bump!

**Comment by guidiamond** [comment]
Bump
